### PR TITLE
connection: fix implementation (by @perfect-daemon)

### PR DESCRIPTION
**Pull request by j-berman**
Cherry-picked the commits from @perfect-daemon's #7760, resolved merge conflicts, and made most of the style changes requested in the PR. I made 0 logic changes to the PR. @perfect-daemon is the author of the PR and deserves full credit for the work.

#7760 is functionally a rewrite of the code handling server-side socket connections. It solves important issues that can be observed by running the tests provided in the PR, with the `monerod` code from before this PR.

Summarizing the issues the 3 tests demonstrate:
1. the SSL shutdown can hang,
2. the SSL handshake sequence can seg fault,
3. there is a circular dependency in the strand processing reads/writes that can lead to unnecessary dropped connections, worse performance, and other problems.

There are additional issues solved by the code beyond what is obvious in the tests. As much as it may not seem like it at first pass, the code does indeed seem to be a good-faith attempt by the author to make the fewest changes possible, as part of the author's goal to make even more structural changes in the future.

Here is a summary of the changes I made in this PR:
- Resolved merge conflicts with master in commit 3be1dbd0.
	- [here is the diff](https://paste.debian.net/1246421/) with @perfect-daemon's 925c0364 that I cherry-picked to make 3be1dbd0. The changes are small.
- Class members now have an `m_` prefix, clearly differentiating them from local variables when referenced (addresses [comment](https://github.com/monero-project/monero/pull/7760#discussion_r884563950)).
- Cleaned up ambiguous `socket_t` and `timer_t` types (addresses [comment](https://github.com/monero-project/monero/pull/7760#discussion_r889463319) and [comment](https://github.com/monero-project/monero/pull/7760#discussion_r889463385)).
- Replaced the `and` operator with `&&`, replaced `not` with `!` (addresses [comment](https://github.com/monero-project/monero/pull/7760#discussion_r884586499)).
- Reduced prevalence of type aliasing (addresses [comment](https://github.com/monero-project/monero/pull/7760#discussion_r693536969)).
- Added a couple comments.
- Other smaller changes.

### CODE REVIEW

**Comment by r4v3r23** [comment]
thanks for this

File: contrib/epee/include/net/abstract_tcp_server2.inl [hunk 1]
```diff
@@ -562,21 +559,16 @@ namespace net_utils
...
+    m_state.socket.wait_shutdown = true;
     auto on_shutdown = [this, self](const ec_t &ec){
-      lock_guard_t guard(state.lock);
-      state.socket.wait_shutdown = false;
-      if (state.socket.cancel_shutdown) {
-        state.socket.cancel_shutdown = false;
-        if (state.status == status_t::RUNNING)
-          interrupt();
-        else if (state.status == status_t::INTERRUPTED)
-          terminate();
```
[the remark this answers is not in the document]
Comment (j-berman) [reply contrib/epee/include/net/abstract_tcp_server2.inl]: Inadvertent change on my part. Good spot. Reverted.

> Probably worth checking whether this is relevant.

If it reaches this section, the socket has been shut down, so calling `terminate()` here isn't outlandish, though it does deviate from the cancellation pattern used elsewhere.

Also worth highlighting: after more digging, I don't even see how to hit the `INTERRUPTED` status in this spot; it doesn't seem possible given the current code. The author's intent with this was to set the code up for future changes (and more concretely, to enable finer grained control to stop/start the connection with async code) -- this section's structure seems in line with that aim, though this line in particular doesn't seem useful as is.

**Comment by nahuhh** [comment]
@rbrunner7 changes look good?

**Review by rbrunner7 — APPROVED** [review]
The only issue that I found is addressed now. In the very narrow sense that I carefully checked that only "mechanical" but no logical changes to @perfect-daemon 's code were done, I can approve this PR.

**Pull request merged**